src/main/java/io/cryostat/platform/internal/KubeApiPlatformClient.java

```diff
@@ -141,63 +155,275 @@ public void onClose(WatcherException watcherException) {
+                    .addChildNode(
+                            new TargetNode(
+                                    KubernetesNodeType.ENDPOINT, targetTuple.toServiceRef()));
+
+            Pair<HasMetadata, EnvironmentNode> node = pod;
+            while (true) {
+                Pair<HasMetadata, EnvironmentNode> owner = getOrCreateOwnerNode(node);
+                if (owner == null) {
+                    break;
+                }
+                EnvironmentNode ownerNode = owner.getRight();
+                ownerNode.addChildNode(node.getRight());
+                node = owner;
+            }
+            nsNode.addChildNode(node.getRight());
+        } else {
+            // if the Endpoint points to something else(?) than a Pod, just add the target straight
+            // to the Namespace
+            nsNode.addChildNode(
+                    new TargetNode(KubernetesNodeType.ENDPOINT, targetTuple.toServiceRef()));
+        }
+    }
+
+    private Pair<HasMetadata, EnvironmentNode> getOrCreateOwnerNode(
+            Pair<HasMetadata, EnvironmentNode> child) {
+        HasMetadata childRef = child.getLeft();
+        if (childRef == null) {
+            logger.error(
+                    "Could not locate node named {} of kind {} while traversing environment",
+                    child.getRight().getName(),
+                    child.getRight().getNodeType());
+            return null;
+        }
+        List<OwnerReference> owners = childRef.getMetadata().getOwnerReferences();
+        // Take first "expected" owner Kind from NodeTypes, or if none, simply use the first owner
+        OwnerReference owner =
+                owners.stream()
+                        .filter(o -> KubernetesNodeType.fromKubernetesKind(o.getKind()) != null)
+                        .findFirst()
+                        .orElse(owners.get(0));
+        return discoveryNodeCache.computeIfAbsent(cacheKey(owner), this::queryForNode);
+    }
+
+    private Pair<String, String> cacheKey(OwnerReference resource) {
+        return Pair.of(resource.getKind(), resource.getName());
+    }
+
+    // Unfortunately, ObjectReference and OwnerReference both independently implement getKind and
+    // getName - they don't come from a common base class.
+    private Pair<String, String> cacheKey(ObjectReference resource) {
+        return Pair.of(resource.getKind(), resource.getName());
+    }
+
+    private Pair<HasMetadata, EnvironmentNode> queryForNode(Pair<String, String> lookupKey) {
+        KubernetesNodeType nodeType = KubernetesNodeType.fromKubernetesKind(lookupKey.getLeft());
+        String nodeName = lookupKey.getRight();
+        if (nodeType == null) {
+            return null;
+        }
+        synchronized (queryLocks.computeIfAbsent(lookupKey, k -> new Object())) {
+            EnvironmentNode node;
+            HasMetadata ownerRef =
+                    nodeType.getQueryFunction().apply(k8sClient).apply(namespace).apply(nodeName);
```

Comment:
Should this be called ownerRef? It's not an `OwnerReference` so I found it a bit confusing.